gemini/agent-engine/tutorial_plangen_on_naturalplan_examples.ipynb

```diff
@@ -0,0 +1,2325 @@
+        "        # Parse reward and ensure it's an integer between -100 and 100\n",
+        "        try:\n",
+        "            reward = int(reward_output.split('Score:')[-1].strip())\n",
+        "            reward = max(-100, min(100, reward))  # Clamp between -100 and 100\n",
+        "        except:\n",
+        "            reward = -100\n",
+        "\n",
+        "        return reward\n",
+        "\n",
+        "    def _complete_(self, node):\n",
+        "        \"\"\"Generate a zero-shot answer.\"\"\"\n",
+        "        complete_prompt = COMPLETE_PROMPT.format(problem=self.problem, par_str=node.par_str)\n",
+        "        complete = call_gemini(complete_prompt, model_name=self.model_name)\n",
+        "        return complete\n",
+        "\n",
+        "    def expand_node(self, node):\n",
+        "        children = []\n",
+        "        for i in range(self.no_samples):\n",
+        "            step_prompt = STEP_PROMPT.format(problem=self.problem, par_str=node.par_str)\n",
+        "            step = call_gemini(step_prompt, model_name=self.model_name, temperature=0.7)\n",
+        "            children.append(Node(step=step, par_str=node.par_str +\"\\n\"+step, parent=node))\n",
+        "\n",
+        "        node.add_children(children)\n",
+        "\n",
+        "    def print_tree(self, node=None, level=0):\n",
+        "        if node is None:\n",
+        "            node = self.root\n",
+        "        print(\"  \" * level + f\"Step: {node.step}, Reward: {node.reward}, Complete: {node.complete}\")\n",
+        "        for child in node.children:\n",
+        "            self.print_tree(child, level + 1)\n",
+        "\n",
+        "    def run(self):\n",
+        "        self.initialize()\n",
+        "        incomplete = True\n",
+        "        current_node = self.root\n",
+        "        max_iterations = 20  # Adjust this value based on your needs\n",
+        "        iteration = 0\n",
+        "        while incomplete and iteration < max_iterations:\n",
+        "            self.eval_children(current_node)\n",
+        "            for child in current_node.children:\n",
+        "                if '1' in str(child.complete):\n",
+        "                    incomplete = False\n",
+        "                    final_answer = \"\"\n",
+        "                    return child.par_str, final_answer\n",
+        "                iteration += 1\n",
+        "\n",
+        "                # If we reach here, return the best child based on reward\n",
+        "                if iteration >= max_iterations:\n",
+        "                    best_child = max(current_node.children, key=lambda x: x.reward)\n",
+        "                    return best_child.par_str, \"Solution reached iteration limit\"\n",
+        "\n",
+        "            current_node = self.select_node(current_node)\n",
+        "            self.expand_node(current_node)"
```

Comment:
Can any of the code in this notebook be simplified or use external libraries? It's easy to get lost in all of the implementations.

You can also try pulling some of these classes out into external `.py` files and import them.